### PR TITLE
JENKINS-66446 Agent reconnects on Java 11 and websocket

### PR DESCRIPTION
**Pull request by timja**
Extracts the logic from https://github.com/jenkinsci/remoting/pull/295 which wasn't re-used in the websocket protocol

See https://issues.jenkins.io/browse/JENKINS-66446

To reproduce:

1. Use Java 11 on the agent
2. Create a Permanent agent with websocket protocol enabled
3. Restart the controller
4. It should fail to reconnect the agent with a `NoClassDefFoundError: jenkins/slaves/restarter/JnlpSlaveRestarterInstaller`
5. Repeat steps with this patch, reconnects just fine

### CODE REVIEW

**Comment by jglick** [comment]
Ought to be possible to reproduce this in a functional test in core I guess?

**Review by jglick — APPROVED** [review]
It is correct to apply the same logic to WS mode as TCP mode. The original fix seems wrong. If you look at the stack trace, it seems that `JnlpSlaveRestarterInstaller` is _supposed_ to be running in this situation but is failing because it relies on remote class loading, and by the time we get here the channel is dead and it is too late to do that. You can see from https://javadoc.jenkins.io/jenkins/slaves/restarter/SlaveRestarter.html#restart-- that code is prepared for this possibility, but it is not working currently. I suspect https://github.com/jenkinsci/jenkins/pull/3333 inadvertently broke this by causing `JnlpSlaveRestarterInstaller` to no longer be loaded on the agent during _connection_ yet still sending over a `EngineListenerAdapter` which requires `JnlpSlaveRestarterInstaller` to be loaded (even if unused) during _disconnection_; and perhaps the Java 11 class linkage system behaves differently here, due to the “nest” system? (The inner class ought to refer only to its parent `static` nested class but apparently on Java 11 it refers to the outermost class as well?)

At any rate, I think this patch is OK, though I would probably log at `WARNING` rather than `FINE`, and the bug is in core.

File: src/main/java/hudson/remoting/Engine.java
Comment (timja) [inline src/main/java/hudson/remoting/Engine.java:700]:
```suggestion
            LOGGER.log(Level.WARNING, "Reconnection error.", e);
```

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by jeffret-b — APPROVED** [review]
Looks like it should help in this situation also. We've seen less reports of the problem of the other scenarios since the earlier change.

**Review by MarkEWaite — APPROVED** [review]

**Pull request merged**
